datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -1053,15 +1077,123 @@ pub async fn from_substrait_rel(
                     return not_impl_err!("Unsupported exchange kind: {exchange_kind:?}");
                 }
             };
-            Ok(LogicalPlan::Repartition(Repartition {
+            let plan = LogicalPlan::Repartition(Repartition {
                 input,
                 partitioning_scheme,
-            }))
+            });
+            apply_emit_kind(exchange.common.as_ref(), plan)
         }
         _ => not_impl_err!("Unsupported RelType: {:?}", rel.rel_type),
     }
 }
 
+fn retrieve_emit_kind(rel_common: Option<&RelCommon>) -> EmitKind {
+    // the default EmitKind is Direct if it is not set explicitly
+    let default = EmitKind::Direct(rel_common::Direct {});
+    rel_common
+        .and_then(|rc| rc.emit_kind.as_ref())
+        .map_or(default, |ek| ek.clone())
+}
+
+fn contains_volatile_expr(proj: &Projection) -> Result<bool> {
```

Comment:
okay, another one :laughing: I have a slight preference for checking for uniqueness rather than volatility. Volatility probably does take care of correctness issues, but what if it's a fairly complicated subquery expression that's being repeated? Uniqueness check feels safer to me as it definitely won't lead to performance issues on some strange edge cases. wdyt?

Reply:
> but what if it's a fairly complicated subquery expression that's being repeated?

I'm going to choose to leave that as a follow-up and/or punt to the optimizer if that's okay?

Reply:
yes, I'm fine with that. We can always revisit if this actually has an impact on something.